torch/testing/_internal/common_modules.py

```diff
@@ -0,0 +1,203 @@
+import torch
+from copy import deepcopy
+from functools import wraps, partial
+from itertools import chain
+from torch.testing import floating_types
+from torch.testing._internal.common_device_type import (
+    _TestParametrizer, _dtype_test_suffix, _update_param_kwargs, skipIf)
+from torch.testing._internal.common_nn import nllloss_reference
+from torch.testing._internal.common_utils import make_tensor
+from typing import List
+
+
+# List of all namespaces containing modules to test.
+MODULE_NAMESPACES = [
+    torch.nn.modules,
+    torch.nn.qat.modules,
+    torch.nn.quantizable.modules,
+    torch.nn.quantized.modules,
+]
+
+# Modules that shouldn't be tested for one reason or another.
+MODULES_TO_SKIP = {
+    torch.nn.Module,  # abstract base class
+    torch.nn.Container,  # deprecated
+    torch.nn.NLLLoss2d,  # deprecated
+    torch.nn.quantized.modules._ConvNd,  # abstract base class
+    torch.nn.quantized.MaxPool2d,  # aliases to nn.MaxPool2d
+}
+
+# List of all module classes to test.
+MODULE_CLASSES = list(chain(*[[getattr(namespace, module_name) for module_name in namespace.__all__]
+                              for namespace in MODULE_NAMESPACES]))
+MODULE_CLASSES = [cls for cls in MODULE_CLASSES if cls not in MODULES_TO_SKIP]
+
+# Dict of module class -> common name. Useful for making test names more intuitive.
+# Example: torch.nn.modules.linear.Linear -> "nn.Linear"
+MODULE_CLASS_NAMES = {}
+for namespace in MODULE_NAMESPACES:
+    for module_name in namespace.__all__:
+        module_cls = getattr(namespace, module_name)
+        namespace_name = namespace.__name__.replace('torch.', '').replace('.modules', '')
+        MODULE_CLASS_NAMES[module_cls] = f'{namespace_name}.{module_name}'
+
+
+class modules(_TestParametrizer):
```

Comment:
Although "common_device_type.py" is an increasingly silly name, we might want this to go next to the ops decorator for clarity. Might as well add a comment, even if it's only one line:

```
PROTOTYPE @modules decorator that instantiates tests for each module
```

?

Reply:
Hm, in an attempt to not increase the silliness of `common_device_type.py` further, I do prefer the `@modules` decorator being here. This is also the most obvious place I'd think to find it

Reply:
Will add the comment!

Reply:
Sure